EIP4907: ERC-721 User And Expires Extension (#4907)
ethereum/EIPs: EIPS/eip-4907.md

---
eip: 4907
title: ERC-721 User And Expires Extension

> description: Standard interface extension for ERC-721 user and expires.

Reviewer: I don't think this description provides any additional information over the title. I might suggest:

```suggestion
description: Add a time-limited role with restricted permissions to ERC-721 tokens
```

author: Anders (@0xanders), Lance (@LanceSnow), Shrug <shrug@emojidao.org>
discussions-to: https://ethereum-magicians.org/t/idea-erc-721-user-and-expires-extension/8572
status: Draft
type: Standards Track
category: ERC
created: 2022-03-11
requires: 165, 721
---

## Abstract

> This standard is an extension of [ERC-721](./eip-721.md). It proposes an additional role **user** and a valid duration indicator **expires**. It allows users and developers manage the use right more simple and efficient.

Reviewer: ```suggestion
This standard is an extension of [ERC-721](./eip-721.md). It proposes an additional role (`user`) which can be granted to addresses, and a time where the role is automatically revoked (`expires`). The `user` role represents permission to "use" the NFT, but not be able to transfer it or set operators. 
```

## Motivation

> Some NFTs have certain utilities. In-game NFTs can be used to play, virtual land can be used to build scenes, music NFT can be used to enjoy, etc. But in some cases, the owner and user may not be the same person. People may invest in an NFT with utility, but they may not have time or ability to use it. So separating use right from ownership makes a lot of sense.

Reviewer: ```suggestion
Some NFTs have certain utilities. For example: in-game NFTs can be "used" to play, virtual land can be "used" to build scenes, and music NFTs can be "used" while listening. In some cases, the owner and user may not be the same account. Someone may purchase an NFT with utility, but they may not have time or ability to use it, so separating the "use" right from ownership makes a lot of sense.
```

> Nowadays, many NFTs are managed by adding the role of **controller/operator**. People in these roles can perform specific usage actions but can’t approve or transfer the NFT like an owner. If owner plans to set someone as **controller/operator** for a certain period of time, owner needs to submit two on-chain transactions, at the start time and the end time.

Reviewer: ```suggestion
Nowadays, many NFTs are managed by adding the role of **controller/operator**. Accounts in these roles can perform specific usage actions but can’t approve or transfer the NFT like an owner. If an owner plans to set someone as **controller/operator** for a certain period of time, owner needs to submit two on-chain transactions, at the start time and the end time.
```

NFTs may be used in multiple applications, and adding the user role to NFTs makes it easier for the application to make special grants of rights.

### Simple On-chain Time Management

Most NFTs do not take into account the expiration time even though the role of the user is added, resulting in the need for the owner to manually submit on-chain transaction to cancel the user rights, which does not allow accurate on-chain management of the use time and will waste gas.

The usage right often corresponds to a specific time, such as deploying scenes on land, renting game props,  etc. Therefore, it can reduce the on-chain transactions and save gas with **expires**.

### Easy Third-Party Integration

The standard makes it easier for third-party protocols to manage NFT usage rights without permission from the NFT issuer or the NFT application.

## Backwards Compatibility

As mentioned in the specifications section, this standard can be fully ERC721 compatible by adding an extension function set.

In addition, new functions introduced in this standard have many similarities with the existing functions in ERC721. This allows developers to easily adopt the standard quickly.

## Test Cases

> Test cases for EIP4907 are at https://github.com/emojidao/EIP4907 .

Reviewer: Please either inline the test cases in this document, or include them in the `assets/eip-4907` directory.

## Reference Implementation

```solidity

> // SPDX-License-Identifier: MIT

Reviewer: ```suggestion
// SPDX-License-Identifier: CC0-1.0
```
